providers/common/kdfs/pbkdf2.c

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright 2018-2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <stdarg.h>
+#include <string.h>
+#include <openssl/hmac.h>
+#include <openssl/evp.h>
+#include <openssl/kdf.h>
+#include <openssl/core_names.h>
+#include "internal/cryptlib.h"
+#include "internal/numbers.h"
+#include "internal/evp_int.h"
+#include "internal/provider_ctx.h"
+#include "internal/providercommonerr.h"
+
+/* Constants specified in SP800-132 */
+#define KDF_PBKDF2_MIN_KEY_LEN_BITS  112
+#define KDF_PBKDF2_MAX_KEY_LEN_DIGEST_RATIO 0xFFFFFFFF
+#define KDF_PBKDF2_MIN_ITERATIONS 1000
+#define KDF_PBKDF2_MIN_SALT_LEN   (128 / 8)
+/*
+ * For backwards compatibility reasons,
+ * Extra checks are done by default in fips mode only.
+ */
+#ifdef FIPS_MODE
+# define KDF_PBKDF2_DEFAULT_CHECKS 1
+#else
+# define KDF_PBKDF2_DEFAULT_CHECKS 0
+#endif /* FIPS_MODE */
+
+static OSSL_OP_kdf_newctx_fn kdf_pbkdf2_new;
+static OSSL_OP_kdf_freectx_fn kdf_pbkdf2_free;
+static OSSL_OP_kdf_reset_fn kdf_pbkdf2_reset;
+static OSSL_OP_kdf_derive_fn kdf_pbkdf2_derive;
+static OSSL_OP_kdf_settable_ctx_params_fn kdf_pbkdf2_settable_ctx_params;
+static OSSL_OP_kdf_set_ctx_params_fn kdf_pbkdf2_set_ctx_params;
+
+static int  pbkdf2_derive(const char *pass, size_t passlen,
+                          const unsigned char *salt, int saltlen, uint64_t iter,
+                          const EVP_MD *digest, unsigned char *key,
+                          size_t keylen, int extra_checks);
+
+typedef struct {
+    void *provctx;
+    unsigned char *pass;
+    size_t pass_len;
+    unsigned char *salt;
+    size_t salt_len;
+    uint64_t iter;
+    const EVP_MD *md;
```

Comment:
You might want to take a closer look at how I did this in the MAC implementations, supporting both fetched and legacy implementations (the latter as a fallback).

Reply:
This is old pre-provider code which was missed in the conversion.  I'll address it.

Reply:
Do both fetched and legacy implementations of underlying digests have to be supported?

The builtin digests are converted to providers, wouldn't this make engines the only source of digests not from providers?

Both MACs and KDFs are new APIs, should they be burdened with legacy support from the beginning?

Reply:
`EVP_MD` is not a new API, so you cannot know that you're rid of legacy.  It's perfectly possible for an *application* to create an `EVP_MD` and add it to the object db with `EVP_add_md()`.  Should they be able to work with our MAC and KDF implementations?  Let's remember that MACs and KDFs did work with legacy MDs as EVP_PKEY methods, and we do provide EVP_PKEY bridges for them, which means that we do support legacy uses still.

Reply:
I think it would make sense for providers to have a utility sub-library that does this for them in one place.

For MACs there is one instance of this.  KDFs have six.  I don't like the thought of this kind of legacy support being replicated so widely.


I'll put something together tomorrow if I get a chance.  Thoughts about `provider/include` and `provider/lib`?

Reply:
I have nothing against a utility sub-library...  however, I don't understand what you say about MACs...

```
: ; grep -F '*alloc_' providers/*/macs/*.c
providers/common/macs/cmac_prov.c:    EVP_CIPHER *alloc_cipher;    /* fetched CMAC cipher */
providers/common/macs/gmac_prov.c:    EVP_CIPHER *alloc_cipher;    /* Fetched cipher */
providers/common/macs/hmac_prov.c:    EVP_MD *alloc_md;            /* fetched digest */
providers/common/macs/kmac_prov.c:    EVP_MD *alloc_md;            /* Fetched digest */
```

Reply:
I searched for `tmpmd`.

Any reason KMAC doesn't support engines?

Reply:
No reason whatsoever, I guess it just slipped our collective mind...

Reply:
Would this fix be okay as a separate PR?
This one is already getting too big to hold all the details in memory...

Reply:
Sure

Reply:
Issue created so this doesn't get lost: #9768